### DISCUSSION TITLE
Console and Exec access within the web GUI

### DISCUSSION THREAD

**Discussion by The0cdguy** [lxdware/lxd-dashboard#76] [Q&A] [answered]
I have LXD Dashboard now working with access to five (5) LXD servers in a cluster.  I can view details about each of the hosts and their containers.  The only thing not working thus far is terminal access from within the GUI.  I get a certificate error whenever I attempt to use Console or Exec.  The exact verbiage is:

> There was an error connecting to your LXD server.
> The LXD server may be using a self-signed certificate.
> Visit https://my-host.my-domain.org:8443/ to accept the certificate if not yet trusted.

I don't understand this message.  I installed the lxdware.crt on each of my hosts and ran

**lxc config trust add lxdware.crt**

as required.  Is there some other undocumented step I need to follow to get the console to work?  By the way, all of my servers have SSH enabled and I routinely access them from my PC, so terminal access works as it should.

Any ideas?

**Comment by matthewalanpenning** [comment, marked as answer]
Yes, this is an expected popup when accessing the console or exec. Let me explain. When you setup the dashboard to connect to your servers (lxc config trust add lxdware.crt) this allows the dashboard to make changes to your lxd server via the Rest API that each server provides. 

The popup you are seeing now with exec/console is letting you know that your web browser is connecting to a secure web socket (wss) connection and the SSL/TLS certificate that your lxd server is providing to your web browser is a self-signed certificate, meaning it is not verified by a third party. This is the default option with LXD servers and is commonly left that way within internal networks. Once you trust the certificate in your browser you should not see that message again, unless you clear your browser certificate cache.

If you are interested in using public SSL/TLS certificates for your servers, there is minimal information listed in the official LXD documentation at https://documentation.ubuntu.com/lxd/en/latest/authentication/#tls-server-certificate

I hope this clears things up for you.
-Matt

**Reply by The0cdguy** [reply]
Thanks, Matthew.  That clears things up.  I didn't realize the sequence of _what was talking to what._

My browser security settings might also be playing a role.  It's locked down pretty tight.

One thing to note, after a few tries at this, I saw a red notice at the top of the dashboard. I didn't write it down, but it was related to the console sessions being opened. When I looked into the Operations tab, I saw a ton of running tasks that could not be canceled. I think they were all my attempts to create a console session.  It was a couple pages long and I couldn't figure out how to clear it out.  In the end, I lost connection to all but one my clustered servers and finally just nuked the dashboard container and started from scratch with a brand new install.

The dashboard is up now, but I think I will avoid the console feature for the time being.  Maybe it is my browser settings, but I don't want to cause any instability.  I appreciate your help.  This app is great.  I ditched Proxmox for a more hands-on approach to my LXD/Ceph cluster and the dashboard is super helpful.  Before I found it, I had thought of coding something up in Python or Javascript, so you saved me a ton of API reading and effort.

**Reply by matthewalanpenning** [reply]
I have a feeling that it is a security setting blocking the web socket connection and those operations you saw were each cancelled websocket connection. When developing that section of code I saw similar operations left behind when the websocket was cut off without a proper disconnect. The only way I could clear those operations up was by restarting the LXD server. I made a note to go back and look at that code to see if there are additional checks I can add.

You mentioned Python, in addition to the LXD Dasbhoard which is PHP based, I have been porting the code to Python and addition some new features. If interested you can find that here: https://github.com/PenningLabs/lxconsole
